post about portal and the light client landscape (#10)
ethereum/portal-blog: pages/posts/light-clients.md

---
title: State of light clients, and how Portal can help
date: 2023/2/01
description: Explaining why is it important to have multiple client implementations.
tag: client diversity, multiclient, Ethereum, archive
author: jmc

---

## Key Takeaways

- Light clients give access to Ethereum data to small devices, making it possible for apps, wallets, mobiles, bridges and IoT devices to interact with Ethereum. The associated trust assumptions depend upon the specific light client, ranging from completely trusting some RPC provider through to verifying both consensus and execution using SNARKs.
- Light consensus clients are currently easier to build than light execution clients because a dedicated light client protocol for consensus clients was launched at The Merge.

> - The Portal Network enables users to get Beacon headers and `eth_getProof` responses from a peer-to-peer network, solving the problem of centralized third party RPC providers.

Reviewer: ```suggestion
- The Portal Network enables  `eth_getProof` responses from a peer-to-peer network solving the problem of lightweight decentralized access to the state.
```

> - The Portal Network also provides a decentralized storage network for historical data, enabling EIP 4444 (history expiry). This allows for lighter full nodes.

Reviewer: ```suggestion
- The Portal Network provides decentralized access to historical block data, enabling EIP 4444 (history expiry), allowing for lighter full nodes.
```

- Full nodes can be made lighter by expiring state data, but this requires multiple difficult protocol upgrades. (non inclusive list is Verkle tries, and address space expansion)
- Proof systems such as SNARKs are already being used to verify sync committees, but they will likely be used to verify execution, such as the state transition function.

> - Further into the future, stateless clients will provide very lightweight access to Ethereum, but there are substantial prerequisites that must be shipped first, including Verkle tries.

Reviewer: I don't think this is accurate.  Stateless clients are actually likely to not provide much of any access to Ethereum data since they will not have a state database...

Author: yea, thanks. This was a clumsy sentence. IIUC stateless clients will be useful because they can verify witnesses and that's enouhg for a validator to run on small hardware, and they might be able to participate in transaction gossiping. Is that about right?

## What is a light client?

A light client is a piece of software that provides access to Ethereum data without storing large amounts locally. This allows users to interact with Ethereum cheaply with minimal hardware requirements.

> This very broad definition encompasses a spectrum of designs ranging from apps that request data from an RPC provider and forward it to the user unchecked, right through to clients that verify proofs demonstrating that incoming data was both correctly computed and attested by a supermajority of Ethereum validators. 

Reviewer: ```suggestion
This very broad definition encompasses a spectrum of designs ranging from apps that request data from an RPC provider and forward it to the user unchecked, right through to clients that use proofs demonstrating strong cryptographic guarantees about the validity of incoming data. 
```

### Where are we today?

The light clients that exist today are primarily consensus light clients that get their data from RPC requests to a full node and verify against state roots embedded in Beacon block headers. Typically the light client verifies the sync committee signatures, but some newer clients verify SNARKs over the sync committee protocol instead, achieving the same thing but with less computation. Either way, these consensus clients require access to execution data to do anything more interesting than tracking the tip of the chain.

Getting this execution data requires trusting a third party RPC provider to provide honest data without censoring or doing anything nefarious with your information. Some RPC providers expose `eth_getProof` so that you can independently verify the data they provide against a state root from a trusted block header, but some don’t. 

> The table below shows which RPC methods are exposed by each of three clients representing a full (execution + Beacon) client, two light clients ([Lodestar’s light client rpc-prover app](https://chainsafe.github.io/lodestar/lightclient-prover/prover/) and [Helios](https://github.com/a16z/helios/blob/master/rpc.md)) and a Portal client (could be Trin, Fluffy or Ultralight - they aim for feature parity). 

Reviewer 2: Shameless plug: Nimbus also also has similar client: https://github.com/status-im/nimbus-eth1/tree/master/nimbus_verified_proxy

It uses an integrated version of the the Nimbus consensus light client.
